Prototype streaming for declarative shadow DOM

rniwa opened the pull request:
#### c0bda7319ffbab5ebf316e8dec18e50afbfde67a
<pre>
Prototype streaming for declarative shadow DOM
<a href="https://bugs.webkit.org/show_bug.cgi?id=245817">https://bugs.webkit.org/show_bug.cgi?id=245817</a>

Reviewed by Chris Dumez.

Add a very experimental internal settings to make declarative shadow DOM stream by attaching shadow root
when the start tag of a template element is parsed instead of when the end tag is parsed as proposed.

Having this internal setting allow us to analyze the performance characteristics of both approaches.

* Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml: Added a new internal settings.
We don&apos;t use experimental features since this feature needs to be disabled by default in WKTR.

* Source/WebCore/html/HTMLTemplateElement.cpp:
(WebCore::HTMLTemplateElement::fragmentForInsertion const): Added. This is an abstraction around
regular template and template element, which is a parser macro for shadow root.
(WebCore::HTMLTemplateElement::content const): Assert that nobody calls this version unless we&apos;re
dealing with a regular template element (i.e. not a declarative shadow DOM).
(WebCore::HTMLTemplateElement::setDeclarativeShadowRoot): Added.
(WebCore::HTMLTemplateElement::attachAsDeclarativeShadowRootIfNeeded): Exit early if we had attached
the shadow root with the start tag.

* Source/WebCore/html/HTMLTemplateElement.h:

* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::insert):
(WebCore::HTMLConstructionSite::insertHTMLTemplateElement): Added. This implements the main logic
of streaming declarative shadow DOM, which attaches shadow root as soon as the start tag is parsed.
(WebCore::HTMLConstructionSite::ownerDocumentForCurrentNode):

* Source/WebCore/html/parser/HTMLConstructionSite.h:

* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::HTMLTreeBuilder::processTemplateStartTag):

Canonical link: <a href="https://commits.webkit.org/255020@main">https://commits.webkit.org/255020@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b2439e85ffaa65d0f8486f1b80de0d28900aafb2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/90978 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/35549 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/21515 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/100322 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/158827 "Reverted pull request changes (failure)") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/94984 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/34050 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/29068 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/83323 "Built successfully") | [💥 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/97116 "An unexpected error occured. Recent messages:Failed to print configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/96634 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-16-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/77765 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/26937 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/81884 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/81612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/69972 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/82482 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/35130 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/15631 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/77483 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/32930 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/16612 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26697 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/36708 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/39580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/80078 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1534 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/38635 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/35701 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17543 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/html/HTMLTemplateElement.cpp:
```diff
@@ -68,13 +69,26 @@ DocumentFragment* HTMLTemplateElement::contentIfAvailable() const
 
+void HTMLTemplateElement::setDeclarativeShadowRoot(ShadowRoot& shadowRoot)
+{
+    m_declarativeShadowRoot = WeakPtr { shadowRoot };
```
nit: Do we really need the `WeakPtr { }`?

rniwa replied:
No needed after all. Will fix.

cdumez reviewed (approved):
LGTM

rniwa committed:
Prototype streaming for declarative shadow DOM
https://bugs.webkit.org/show_bug.cgi?id=245817

Reviewed by Chris Dumez.

Add a very experimental internal settings to make declarative shadow DOM stream by attaching shadow root
when the start tag of a template element is parsed instead of when the end tag is parsed as proposed.

Having this internal setting allow us to analyze the performance characteristics of both approaches.

* Source/WTF/Scripts/Preferences/WebPreferencesInternal.yaml: Added a new internal settings.
We don't use experimental features since this feature needs to be disabled by default in WKTR.

* Source/WebCore/html/HTMLTemplateElement.cpp:
(WebCore::HTMLTemplateElement::fragmentForInsertion const): Added. This is an abstraction around
regular template and template element, which is a parser macro for shadow root.
(WebCore::HTMLTemplateElement::content const): Assert that nobody calls this version unless we're
dealing with a regular template element (i.e. not a declarative shadow DOM).
(WebCore::HTMLTemplateElement::setDeclarativeShadowRoot): Added.
(WebCore::HTMLTemplateElement::attachAsDeclarativeShadowRootIfNeeded): Exit early if we had attached
the shadow root with the start tag.

* Source/WebCore/html/HTMLTemplateElement.h:

* Source/WebCore/html/parser/HTMLConstructionSite.cpp:
(WebCore::insert):
(WebCore::HTMLConstructionSite::insertHTMLTemplateElement): Added. This implements the main logic
of streaming declarative shadow DOM, which attaches shadow root as soon as the start tag is parsed.
(WebCore::HTMLConstructionSite::ownerDocumentForCurrentNode):

* Source/WebCore/html/parser/HTMLConstructionSite.h:

* Source/WebCore/html/parser/HTMLTreeBuilder.cpp:
(WebCore::HTMLTreeBuilder::processTemplateStartTag):

Canonical link: https://commits.webkit.org/255020@main
